# Homework

## Dataset
In this homework, we'll build a model for predicting if we have an image of a dog or a cat. For this, we will use the "Dogs & Cats" dataset that can be downloaded from Kaggle.

You need to download the train.zip file.

If you have troubles downloading from Kaggle, use this link instead:

wget https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip

In the lectures we saw how to use a pre-trained neural network. In the homework, we'll train a much smaller model from scratch.

Note: You don't need a computer with a GPU for this homework. A laptop or any personal computer should be sufficient.

## Data Preparation

The dataset contains 12,500 images of cats and 12,500 images of dogs.

Now we need to split this data into train and validation

Create a train and validation folders
In each folder, create cats and dogs folders
Move the first 10,000 images to the train folder (from 0 to 9999) for boths cats and dogs - and put them in respective folders

Move the remaining 2,500 images to the validation folder (from 10000 to 12499)
You can do this manually or with Python (check os and shutil packages).



In [1]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip -O train.zip

--2021-11-21 13:41:58--  https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/426348925/f39169c9-5f22-4a57-bb37-495c0d2974ab?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211121%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211121T124158Z&X-Amz-Expires=300&X-Amz-Signature=4b29fa56afa8ec52657660ed9bac52790edc889cd74acf70eb54bf47728ef585&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=426348925&response-content-disposition=attachment%3B%20filename%3Dtrain.zip&response-content-type=application%2Foctet-stream [following]
--2021-11-21 13:41:58--  https://github-releases.githubusercontent.com/426348925/f39169c9-5f22-4a57-bb37-495c0d2974ab?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVE

In [2]:
!unzip -qo 'train.zip' -d 'data'

In [3]:
!mkdir 'train'
!mkdir 'train/cats'
!mkdir 'train/dogs'
!mkdir 'validation'
!mkdir 'validation/cats'
!mkdir 'validation/dogs'

In [4]:
!bash -c 'mv data/train/cat.{0..9999}.jpg train/cats'
!bash -c 'mv data/train/dog.{0..9999}.jpg train/dogs'

In [5]:
!bash -c 'mv data/train/cat.{10000..12499}.jpg validation/cats'
!bash -c 'mv data/train/dog.{10000..12499}.jpg validation/dogs'

In [6]:
!rm -drf data

## Model
For this homework we will use Convolutional Neural Network (CNN. Like in the lectures, we'll use Keras.

You need to develop the model with following structure:

* The shape for input should be (150, 150, 3)
* Next, create a covolutional layer (Conv2D):
    * Use 32 filters
    * Kernel size should be (3, 3) (that's the size of the filter)
    * Use 'relu' as activation
* Reduce the size of the feature map with max pooling (MaxPooling2D)
    * Set the pooling size to (2, 2)
* Turn the multi-dimensional result into vectors using a Flatten layer
* Next, add a Dense layer with 64 neurons and 'relu' activation
* Finally, create the Dense layer with 1 neuron - this will be the output

As optimizer use SGD with the following parameters:

* SGD(lr=0.002, momentum=0.8)


In [7]:
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

import tensorflow as tf
from tensorflow import keras

In [8]:
def make_model():
    inputs = keras.layers.Input(shape=(150,150,3))

    conv = keras.layers.Conv2D(filters=32, kernel_size=(3,3), activation='relu')(inputs)

    pool = keras.layers.MaxPooling2D(pool_size=(2,2))(conv)

    vectors = keras.layers.Flatten()(pool)

    dense = keras.layers.Dense(64, activation='relu')(vectors)

    outputs = keras.layers.Dense(1,activation='sigmoid')(dense)

    model = keras.Model(inputs, outputs)

    optimizer = keras.optimizers.SGD(learning_rate=0.002, momentum=0.8)
    loss = keras.losses.BinaryCrossentropy(from_logits=False)
    
    model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])    
    
    return model


## Question 1
__Since we have a binary classification problem, what is the best loss function for us?__

__*Answer Q1: BinaryCrossentropy*__


## Question 2

__What's the total number of parameters of the model? You can use the summary method for that.__

In [9]:
model = make_model()

model.summary()

Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 150, 150, 3)]     0         
_________________________________________________________________
conv2d (Conv2D)              (None, 148, 148, 32)      896       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 74, 74, 32)        0         
_________________________________________________________________
flatten (Flatten)            (None, 175232)            0         
_________________________________________________________________
dense (Dense)                (None, 64)                11214912  
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 65        
Total params: 11,215,873
Trainable params: 11,215,873
Non-trainable params: 0
_________________________________________________

2021-11-21 13:43:14.444809: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2021-11-21 13:43:14.444938: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
2021-11-21 13:43:14.445013: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (ubuntu): /proc/driver/nvidia/version does not exist
2021-11-21 13:43:14.466647: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


__*Answer Q2: 11,215,873*__

## Generators and Training
For the next two questions, use the following data generator for both train and validation:

```python
ImageDataGenerator(rescale=1./255)
```

We don't need to do any additional pre-processing for the images.

For training use .fit() with the following params:

```python
model.fit(
    train_generator,
    steps_per_epoch=100,
    epochs=10,
    validation_data=validation_generator,
    validation_steps=50
)
```

In [10]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [11]:
train_generator = ImageDataGenerator(rescale=1./255)

train_ds = train_generator.flow_from_directory(
    './train',
    target_size=(150,150),
  #  classes=['dogs','cats'],
    class_mode='binary',
    batch_size=20,
)

validation_generator = ImageDataGenerator(rescale=1./255)

val_ds = validation_generator.flow_from_directory(
    './validation',
    target_size=(150,150),
 #   classes=['dogs','cats'],
    class_mode='binary',
    batch_size=20,
    shuffle=False
)

Found 20000 images belonging to 2 classes.
Found 5000 images belonging to 2 classes.


In [12]:
train_ds.class_indices, val_ds.class_indices

({'cats': 0, 'dogs': 1}, {'cats': 0, 'dogs': 1})

In [13]:
history = model.fit(
    train_ds,
    steps_per_epoch=100,
    epochs=10,
    validation_data=val_ds,
    validation_steps=50
)

2021-11-21 13:43:15.268936: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)


Epoch 1/10
  1/100 [..............................] - ETA: 49s - loss: 0.6879 - accuracy: 0.5500

2021-11-21 13:43:15.595890: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 56074240 exceeds 10% of free system memory.
2021-11-21 13:43:15.696212: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 56074240 exceeds 10% of free system memory.
2021-11-21 13:43:15.777825: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 56074240 exceeds 10% of free system memory.


  3/100 [..............................] - ETA: 11s - loss: 0.6965 - accuracy: 0.5000

2021-11-21 13:43:15.829969: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 56074240 exceeds 10% of free system memory.
2021-11-21 13:43:15.891345: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 56074240 exceeds 10% of free system memory.


100/100 [==============================] - 13s 130ms/step - loss: 0.6986 - accuracy: 0.4790 - val_loss: 0.6970 - val_accuracy: 0.0090
Epoch 2/10
100/100 [==============================] - 13s 126ms/step - loss: 0.6937 - accuracy: 0.4875 - val_loss: 0.6979 - val_accuracy: 0.1810
Epoch 3/10
100/100 [==============================] - 13s 130ms/step - loss: 0.6929 - accuracy: 0.5095 - val_loss: 0.6981 - val_accuracy: 0.1250
Epoch 4/10
100/100 [==============================] - 12s 123ms/step - loss: 0.6926 - accuracy: 0.5290 - val_loss: 0.7219 - val_accuracy: 0.0180
Epoch 5/10
100/100 [==============================] - 12s 123ms/step - loss: 0.6904 - accuracy: 0.5350 - val_loss: 0.6781 - val_accuracy: 0.5970
Epoch 6/10
100/100 [==============================] - 13s 128ms/step - loss: 0.6849 - accuracy: 0.5600 - val_loss: 0.7816 - val_accuracy: 0.1160
Epoch 7/10
100/100 [==============================] - 13s 125ms/step - loss: 0.6723 - accuracy: 0.5830 - val_loss: 0.7086 - val_accuracy: 0.4

## Question 3  

__What is the median of training accuracy for this model?__

In [14]:
np.median(history.history['accuracy'])

0.5475000143051147

__*Answer Q3: 0.55*__  

## Question 4

__What is the standard deviation of training loss for this model?__

In [15]:
np.std(history.history['loss'])

0.012723315232945192

__*Answer Q4: 0.012*__

## Data Augmentation

For the next two questions, we'll generate more data using data augmentations.

Add the following augmentations to your training data generator:

* rotation_range=40,
* width_shift_range=0.2,
* height_shift_range=0.2,
* shear_range=0.2,
* zoom_range=0.2,
* horizontal_flip=True,
* fill_mode='nearest'

In [16]:
train_generator = ImageDataGenerator(
    rescale=1./255,    
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest')

train_ds = train_generator.flow_from_directory(
    './train',
    target_size=(150,150),
    class_mode='binary',
    batch_size=20,
)

validation_generator = ImageDataGenerator(rescale=1./255)

val_ds = validation_generator.flow_from_directory(
    './validation',
    target_size=(150,150),
    class_mode='binary',
    batch_size=20,
    shuffle=False
)

Found 20000 images belonging to 2 classes.
Found 5000 images belonging to 2 classes.


## Question 5
Let's train our model for 10 more epochs using the same code as previously. Make sure you don't re-create the model - we want to continue training the model we already started training.

__What is the mean of validation loss for the model trained with augmentations?__

In [17]:
history = model.fit(
    train_ds,
    steps_per_epoch=100,
    epochs=10,
    validation_data=val_ds,
    validation_steps=50
)

Epoch 1/10
100/100 [==============================] - 14s 144ms/step - loss: 0.6728 - accuracy: 0.5765 - val_loss: 0.6154 - val_accuracy: 0.6600
Epoch 2/10
100/100 [==============================] - 15s 151ms/step - loss: 0.6814 - accuracy: 0.5555 - val_loss: 0.7038 - val_accuracy: 0.4510
Epoch 3/10
100/100 [==============================] - 17s 165ms/step - loss: 0.6739 - accuracy: 0.5670 - val_loss: 0.7731 - val_accuracy: 0.2850
Epoch 4/10
100/100 [==============================] - 15s 152ms/step - loss: 0.6667 - accuracy: 0.5880 - val_loss: 0.5731 - val_accuracy: 0.7850
Epoch 5/10
100/100 [==============================] - 14s 142ms/step - loss: 0.6747 - accuracy: 0.5580 - val_loss: 0.6275 - val_accuracy: 0.7350
Epoch 6/10
100/100 [==============================] - 14s 142ms/step - loss: 0.6633 - accuracy: 0.5965 - val_loss: 0.5363 - val_accuracy: 0.8390
Epoch 7/10
100/100 [==============================] - 14s 139ms/step - loss: 0.6683 - accuracy: 0.5785 - val_loss: 0.9571 - val_ac

In [20]:
np.mean(history.history['val_loss'])

0.6486403346061707

__*Answer Q5: 0.65*__

## Question 6

__What's the average of validation accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?__

In [21]:
np.mean(history.history['val_accuracy'][-5:])

0.6717999964952469

__*Answer Q6: 0.67*__